### ISSUE TITLE
GBT fails consistently after some time.  CreateNewBlock: TestBlockValidity failed: (code 0)

### ISSUE THREAD

**Issue by flound1129** [namecoin/namecoin-core#50]
$ namecoin-cli getblocktemplate
error code: -1
error message:
CreateNewBlock: TestBlockValidity failed: (code 0)

Getting lots of these in the debug log:
2015-12-23 06:35:18 CreateNewBlock(): total size 1246 txs: 1 fees: 500000 sigops 101
2015-12-23 06:35:18 ERROR: CheckNameTransaction: NAME_FIRSTUPDATE on an unexpired name
2015-12-23 06:35:18 ERROR: CheckInputs: tx invalid for Namecoin
2015-12-23 06:35:18 ERROR: ConnectBlock(): CheckInputs on fc56b9f2297a299cf59022317b50ba45ace9392932453eb027f3a77bf4c90bb1 failed with (code 0)

**Comment by domob1812** [comment]
This seems like a good indication of what the issue may be - thanks a lot.  I have a clue now what to check and possibly fix - although I think that I may only have time to do that after Christmas.

If someone wants to help further, you can try to find a reproducible test case on regtest mode - that would be great, but I think I may be able to fix it even without (and come up with one myself).

**Comment by domob1812** [comment]
This was easier than I thought - testing a fix right now and will submit a pull request once this is done.  In the mean time, let me briefly describe what the issue is:  This arises due to a combination of two recent changes in Bitcoin.

One concerned the mempool logic, which now uses a "dummy coinview" for testing new transactions before accepting into the mempool.  This requires to explicitly load all necessary UTXO information into the dummy cache beforehand and was (I guess) done to improve performance.  This also requires to explicitly load information about the name database.  Part of that was done when I merged, but I forgot one part - this allowed to get registrations of existing names (which are invalid) into the mempool if the RPC code is "patched".

These transactions are, of course, invalid in a block and cannot be mined.  The second change, now, was a rewrite of block creation in the miner.  Previously, all transactions were individually checked against the created block before adding them, so that these invalid transactions would simply not make it into a block.  According to the rewrite, this check is now gone and only ad-hoc tests are done for each transaction - again in the name of performance, I guess.  Also here, I added some checks for Namecoin when merging, but they were not sufficient for the invalid firstupdates (as they are not supposed to get into the mempool at all).

All in all, this leads to the effect that the miner might create invalid blocks if the mempool contains these invalid transactions.  This is not a "network-wide" problem, as the blocks are recognised as invalid as they should.  But this makes the miner fail, which is the issue here.  The fix is easy, we simply have to fix the mempool logic to recognise the transactions in question here.

**Comment by domob1812** [comment]
See https://github.com/namecoin/namecoin-core/pull/51 and, of you can, please try if this fixes the issue for you.

**Comment by flound1129** [comment]
Doesn't look like that PR is passing tests, is that a problem?

**Comment by JeremyRand** [comment]
@flound1129 The OSX failure is due to an issue with upstream Bitcoin Core's dependency server, and only affects Travis CI (not end users).  The two Linux failures are both on proxy_test.py; I have no idea what impact they would have on end users.  Might be a good idea to wait for @domob1812 to look at the proxy_test.py errors before using it for something as mission-critical as mining, but it doesn't appear to be an error relating to mining or consensus code, so I'd guess that it's safe -- but Daniel's the expert on this.

**Comment by domob1812** [comment]
I've not yet fully investigated the recent failures, but I'm sure it is not related to this change.  I also run my own "critical" tests for each commit (including the tests of all the name-specific features), so the code "should" work fine nevertheless.

**Comment by bitkevin** [comment]
hey, I just put merged mining in [btcpool](https://github.com/btccom/btcpool) and I have met the same issue:

* version: `v0.13.0rc1`
* os: `Ubuntu 16.04 LTS 64 bits`

```
$ namecoin-cli getauxblock
error code: -1
error message:
CreateNewBlock: TestBlockValidity failed:  (code 0)
```

```
2016-09-08 11:36:58 CreateNewBlock(): total size 43124 txs: 58 fees: 46700000 sigops 860
2016-09-08 11:36:58 ERROR: CheckNameTransaction: NAME_NEW is not mature for FIRST_UPDATE
2016-09-08 11:36:58 ERROR: CheckInputs: tx invalid for Namecoin
2016-09-08 11:36:58 ERROR: ConnectBlock(): CheckInputs on d8980bdf31ca5ad42721c7385384b6184ea70845a18a3f161f00f0e223975102 failed with  (code 0)
2016-09-08 11:36:58 keypool return 2
```

**Comment by bitkevin** [comment]
I have to add this shell script to make sure namecoind is works fine. More details in this [commit](https://github.com/btccom/btcpool/commit/96c68415bbcb5f88c3d9f60c2d5a0523eb7d73a8).

```
#! /bin/bash
#
# namecoind getauxblock watcher shell
#
# this crontab shell is try to fix issue:
#   https://github.com/namecoin/namecoin-core/issues/50#issuecomment-245571582
#
# @copyright btc.com
# @author Zhibiao Pan
# @since 2016-09
#
SROOT=$(cd $(dirname "$0"); pwd)
cd $SROOT

NAMECOIND_RPC="namecoin-cli "

SUCCESS=`$NAMECOIND_RPC getauxblock | grep "previousblockhash" | wc -l`

if [[ $SUCCESS -ne 1 ]]; then
  # stop namecoind, supervisor will start it again
  echo "got error, restart namecoind..."
  $NAMECOIND_RPC stop
fi
```

**Comment by daizisheng** [comment]
I met this problem too and your way works, thanks @bitkevin

**Comment by crackfoo** [comment]
I'm still getting this problem even on the 0.14 branch...  I'm not sure what the cause would be... is it from improper calling from my pool? Is that possible???

**Comment by JeremyRand** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

@domob1812 Any ideas on this?

-----BEGIN PGP SIGNATURE-----

iQIcBAEBCgAGBQJZNTJAAAoJELPy0WV4bWVwc1EP/1vPVvSVDqVeNr/W8vjghtxB
iLUUIk+edy85RXCQCqbLnsj9rq8FJX7+vNgV7cz3lF8oYEpVUdvKf+WCJJ5N3ovP
h0mLgWOFMdsMXXGhwhN2x9ME/KkQp6L32a+wVQpYs2BcwLXtgKDxG3KP6ry7BkdD
VUMm7gQ6vHR3OV7YgJFcgbk8O4UHoE4BK2P+Ha5M6HoZm5hKDde+J8sMWruXh1vo
j5QLpGNqQaZBflkf8eowurGb8Abh9N7Pn6U6guf7u3SLOqNrQsb9drDF0x2+q9XL
R4NAdoDrYecrbhCNX6MAr7PC6tQnw6Y85hwgdFB+Y3vgmjYWOk0jjHV8mRCvTTkS
U7JVQfX5LzPiXFMSDsuX3JqyEhFsvXatwEaxXgw1+bvzJBLdsgUhKVUpsBcx+R0t
FSn824xb/NoiT1IwXVqk35v9g8FlSqHjoEKlzM+oRKt0sIGfPpWXZvgC7te62dPq
0bApZGoEAh2Cb4OMq/SxLlO74KSQEMENMHMb8pL4PoPW7Cp2lGqmjHALyM3YAV8L
izuDIRt9WqK7LMZc0xo0eiF1Y4C5AxCqi6ZXxrxPXaCWKzT76TLs50RYv+a+BKCJ
s+/W0Y1jR5Z8VI7NM8oPZi2daHHuS8Bn9P/yav7IdUDDmt2GpYcB0GIvctekARlK
Ea+WwApVBcy+52AxTXQx
=kIyf
-----END PGP SIGNATURE-----

**Comment by domob1812** [comment]
@bitkevin, @crackfoo, @daizisheng:  Is this still happening?

**Comment by crackfoo** [comment]
yup, still seeing this now and then... unsure why. will find a block or two then get that.

**Comment by domob1812** [comment]
Thanks, let me know if you find information to reproduce this for debugging.

**Comment by YihaoPeng** [comment]
I met the issue today with the old version: https://github.com/namecoin/namecoin-core/commit/bd72abbb221a32cfa9ab64f577f6b339a135f5c3
When I upgrade the node to the lastest version https://github.com/namecoin/namecoin-core/commit/570a54769056680b1204194fd9a4de94852c8fe7, the issue still exists.

So I stopped the node, deleted `mempool.dat`, and restart it. Then the issue disappeared.

**Comment by domob1812** [comment]
@YihaoPeng, thanks!

This is interesting - the offending transaction `6a09b86047f70ae148b03f7fb761ec6caf65325e99dae368d9489c85a45435ce` is a name_firstupdate, whose name_new is also unconfirmed (`16ba9e81c0cb48510a2f587851e10eff4aecda4fa7019c3f9b2cf7784fa70a22`).  This is something you cannot do with the RPC interface, but you can of course do if you construct the transactions somehow else.  Presumably the logic to prevent adding name_firstupdates for immature name_news to a new block is broken in this edge case - I will take a look.

(And while those transactions are in the mempool, I can indeed reproduce the issue on my node when calling `getblocktemplate`.)

**Comment by crackfoo** [comment]
Yes I’ve gotten this error 5-6 times today...

**Comment by JeremyRand** [comment]
Maybe this bug is the explanation for most of the pools dropping out 2 days ago.  https://github.com/namecoin/namecoin-core/issues/227

**Comment by crackfoo** [comment]
sounds plausible. Is there any way using the cli to invalidate that tx?

**Comment by JeremyRand** [comment]
> Is there any way using the cli to invalidate that tx?

I'm not sure.  @domob1812 would probably know.  (I'll look around and see if I can find the answer in case Domob doesn't get back to us quickly on it.)

**Comment by JeremyRand** [comment]
@crackfoo Maybe the `prioritisetransaction` RPC call could be abused to simulate that transaction as having paid too low a fee, so that it won't be included in mined blocks?  I'm not certain that this will work, but maybe worth a try.  CC: @domob1812

**Comment by JeremyRand** [comment]
Note that the "right" solution is to fix the code rather than try to exclude specific transactions.  But excluding specific transactions might be a decent stopgap while we're waiting for a code fix.

**Comment by domob1812** [comment]
Indeed, this sounds like a plausible explanation for #227!

I'm not aware of an existing RPC to ban the transaction from the mempool.  `prioritisetransaction` can be tried as mentioned by Jeremy, but that needs some fiddling at best.  (I think I used it for the purpose of not mining transactions in regtests for Huntercoin some time ago, but I don't think it is very reliable.)

**Comment by domob1812** [comment]
Now that I can reproduce this, I'll hopefully be able to fix the code tomorrow.  In the mean time, regular restarting of the daemon might help.  (Or a hack could be done in the daemon to blacklist the specific transactions from going into the mempool.)

**Comment by JeremyRand** [comment]
@crackfoo if you try the `prioritisetransaction` trick, please let us know whether it solves the immediate issue for you.  Would be excellent to see whether that can hold us over until there's a code fix.

@domob1812 would it also be a good idea to positively prioritise the name_new transaction, so that it gets confirmed quickly, at which point the immediate issue should solve itself?

**Comment by domob1812** [comment]
I've not looked into why the name_new is unconfirmed, but yes, perhaps you could do that.  However, I guess that the reason might be that creating the block simply fails when you have both transactions, and so mining just the one is not easy without *deprioritising* the other one.

**Comment by JeremyRand** [comment]
@domob1812 right, I meant deprioritise the name_firstupdate and positively prioritise the name_new.

**Comment by domob1812** [comment]
A hack to work around the issue would be to mark the tx as invalid at some place in the code.  For instance in `src/names/main.cpp` inside `CheckNameTransaction`:  Just compare the tx hash and if it matches the name_firstupdate, return false.  This could be a good work around until I can fix the bug tomorrow.

**Comment by JeremyRand** [comment]
> A hack to work around the issue would be to mark the tx as invalid at some place in the code. For instance in src/names/main.cpp inside CheckNameTransaction: Just compare the tx hash and if it matches the name_firstupdate, return false. This could be a good work around until I can fix the bug tomorrow.

@domob1812 hmm, would that affect the consensus rules?

**Comment by domob1812** [comment]
Ah yes, indeed, it would cause issues afterwards.  (But could be used to just mine the name_new.)  Perhaps `src/policy/policy.cpp` and then `IsStandard` is better.

**Comment by domob1812** [comment]
Or just `validation.cpp`'s `AcceptToMemoryPool`.

**Comment by JeremyRand** [comment]
Yeah, the standardness check and mempool should be safe, since they're not consensus-related.  It's probably worth trying `prioritisetransaction` first though, since that doesn't need a recompile and there's less that can go wrong.

**Comment by domob1812** [comment]
Agreed.  I'll go offline for today soon - but will (hopefully) fix the bug correctly tomorrow.  In the mean time, try `prioritisetransaction` first and then hacking `AcceptToMemoryPool`.  As soon as someone manages to get the transaction mined we will be good.  (Unless someone creates another such tx pair as an attack - but also that will be fixed tomorrow.)

**Comment by domob1812** [comment]
Completely untested potential fix for the underlying issue: https://github.com/domob1812/namecore/commit/bcb4f33cd51d91482671970d201895564e634fa8.

I'll redo this tomorrow with proper testing and perhaps in a different way (after thinking about it), but it may (or may not) fix the issue for now if anyone wants to try.

**Comment by JeremyRand** [comment]
(Obviously this fix is a stopgap, and once a proper code fix is merged to Namecoin Core, this stopgap will be unnecessary.)

**Comment by YihaoPeng** [comment]
To fix the issue, simply don't adding the immature `name_firstupdate` when generating a new block (`getblocktemplate` of `createauxblock`). I think there is no need to avoiding it be added to the memory pool.

This is like a transaction with `locktime`, it can be added to the memory pool, but cannot enter the block immediately.

After reaching the height of the specified of `locktime`, the transaction is automatically added to the block. Users do not need to wait for this and re-broadcast it.

**Comment by JeremyRand** [comment]
@domob1812 IMO it's a bad idea for Namecoin Core to allow relaying of `name_firstupdate` transactions that are spending immature `name_new` outputs.  Relaying such transactions allows any third party on the Namecoin network to trivially frontrun the name registration, since it reveals the name that's being registered.  I strongly suspect that anyone who's creating and broadcasting such transactions is doing so accidentally.  So unless there's a use case that I'm missing, my vote is that Namecoin Core should reject such transactions from the mempool and not relay them.

**Comment by domob1812** [comment]
One thing to note:  As soon as the offending transaction gets mined (by one miner being successful with `prioritisetransaction`), the issue will be resolved.  There will not be a need for any other miner to perform that "fix" as well.

@JeremyRand: I think it is perfectly fine to broadcast a `name_firstupdate` after the `name_new` has a few but not necessarily 12 confirmations - this just slightly weakens the security, as it reduces the number of confirmations you have in advance of anyone trying to front run.  But 12 are super cautious and you might choose to be fine with less.  The use would be not having to wait for 12 confirmations before you go offline, for instance, if you are fine with taking the small risk.

Of course, broadcasting a `name_firstupdate` with an *unconfirmed* `name_new` as was the case here is dangerous.  But note that this is something that the current standard RPC interface doesn't allow already (somewhat by accident, but still).  So I think as we are good as long as we fix the mining issue (which I will do now).

**Comment by YihaoPeng** [comment]
@JeremyRand Yes, you're right. This can indeed lead to cybersquatting. I did not realize this problem before.

**Comment by domob1812** [comment]
The proposed fix is in #228.  It is very similar to my "untested fix" from above, but now includes proper regtests (which did reproduce the original issue before applying the code fix).  The "untested fix" also broke the use of unconfirmed *currency* inputs, for instance from multiple `name_firstupdate`'s in a wallet with only one coin output.  This is fixed (and verified in the regtest) now as well.

**Comment by JeremyRand** [comment]
@crackfoo may I ask whether you're a solo miner or a pool operator?  If you're a pool operator, may I ask which pool?  (I'd like to credit the people and pools who were involved in helping us get the issue fixed.)

**Comment by crackfoo** [comment]
I'm a pool op (http://zpool.ca), and the workarounds don't last for long before the error comes up again but I've not tried the #228 fixed yet. Shall I try that now for testing?

**Comment by JeremyRand** [comment]
@crackfoo yes, it would be awesome if you could test https://github.com/namecoin/namecoin-core/pull/228 .  (That said, I'd expect the issue to be resolved anyway since the offending transactions are mined by now... unless there are other transactions in the mempool that also trigger the issue?)

**Comment by crackfoo** [comment]
There must be another as I’m getting alerts from my server again.
I’ll try the code fix later tonight.

Thanks

**Comment by JeremyRand** [comment]
@crackfoo any test results?

**Comment by crackfoo** [comment]
yeah i think I must have messed something up.

keeps warning me about invalid .conf options.

Error reading configuration file: Invalid configuration value enableaccounts

Error reading configuration file: Invalid configuration value rpctimeout


etc... I'll try again.

**Comment by crackfoo** [comment]
ok, got issues sorted.

Running now with that test code. Will let you know if the error pops up again.

**Comment by domob1812** [comment]
This should be fixed now with #228.  Feel free to reopen if not, or to open new issue if different problems occur in the future.
